### PR TITLE
dragonball: Support vhost-user-net device

### PR DESCRIPTION
**Pull request by justxuewei**
This PR introduces vhost-user-net devices to Dragonball. The devices are
allowed to run as server on the VMM side.

Fixes: #8502

Signed-off-by: Eric Ren <renzhen@linux.alibaba.com>
Signed-off-by: Liu Jiang <gerry@linux.alibaba.com>
Signed-off-by: Zha Bin <zhabin@linux.alibaba.com>
Signed-off-by: Chao Wu <chaowu@linux.alibaba.com>
Signed-off-by: Zizheng Bian <zizheng.bian@linux.alibaba.com>
Signed-off-by: Xuewei Niu <niuxuewei.nxw@antgroup.com>

### CODE REVIEW

**Comment by justxuewei** [comment]
Hi @studychao @ZizhengBian @lifupan, vhost-user-net device has been tested with DPDK. Please take a look at this. Thanks!

File: src/dragonball/src/api/v1/vmm_action.rs [hunk 1]
```diff
@@ -712,6 +733,30 @@ impl VmmService {
...
+        config: VhostUserNetDeviceConfigInfo,
+    ) -> VmmRequestResult {
+        let vm = vmm.get_vm_mut().ok_or(VmmActionError::InvalidVMID)?;
+        let ctx = vm
+            .create_device_op_context(Some(event_mgr.epoll_manager()))
+            .map_err(|err| {
+                if let StartMicroVmError::MicroVMAlreadyRunning = err {
+                    VmmActionError::VhostUserNet(VhostUserNetDeviceError::UpdateNotAllowedPostBoot)
+                } else if let StartMicroVmError::UpcallServerNotReady = err {
+                    VmmActionError::UpcallServerNotReady
```
Comment (liubogithub) [inline src/dragonball/src/api/v1/vmm_action.rs:750]: Is upcall ready to use?

File: src/dragonball/src/dbs_virtio_devices/src/vhost/net.rs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+use crate::{DbsGuestAddressSpace, Error as VirtioError, Result as VirtioResult};
+
+pub(crate) trait FromNetCtrl<T> {
+    fn from_net_ctrl_st<M: GuestMemory>(mem: &M, desc: &Descriptor) -> VirtioResult<T> {
+        let mut buf = vec![0u8; std::mem::size_of::<T>()];
+        match mem.read_slice(&mut buf, desc.addr()) {
+            Ok(_) => unsafe { Ok(std::ptr::read_volatile(&buf[..] as *const _ as *const T)) },
+            Err(err) => {
+                error!("Failed to read from memory, {}", err);
```
Comment (liubogithub) [inline src/dragonball/src/dbs_virtio_devices/src/vhost/net.rs:16]: Is the error msg here too generic?  It's difficult to tell it's vhost-net related.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/connection.rs [hunk 3]
```diff
@@ -189,6 +192,9 @@ impl Endpoint {
         Ok(Some(features))
     }
 
+    /// TODO: Mark this function as unused as it is called by
+    /// vhost-user-fd only.
```
Comment (liubogithub) [inline src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/connection.rs:196]: what is vhost-user-fd? typo?

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs [hunk 4]
```diff
@@ -0,0 +1,776 @@
...
+        handler: &VhostUserNetHandler<AS, Q, R>,
+    ) -> ActivateResult
+    where
+        AS: DbsGuestAddressSpace,
+        Q: QueueT + Send + 'static,
+        R: GuestMemoryRegion + Sync + Send + 'static,
+    {
+        trace!(target: "vhost-net", "{}: VhostUserNetDevice::activate_slave()", self.id);
+        // Safe because the `intr_group` has been initialized in
+        // VirtioDevice::activate().
```
Comment (liubogithub) [inline src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:180]: Guess that this is stale comments.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs [hunk 5]
```diff
@@ -0,0 +1,776 @@
...
+            }
+            // Do reconnect
+            // Wait 100ms for the next connection
+            let delay = Duration::from_millis(100);
+            std::thread::sleep(delay);
+            // The underlying communication channel has been disconnected,
+            // recreate it again.
+            let (master, avail_features) = self.listener.accept()?;
+            if !avail_features & self.device_info.acked_features() != 0 {
+                error!("{}: Virtio features changed when reconnecting.", self.id);
```
Comment (liubogithub) [inline src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:223]: Would be helpful to know which features mismatched.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs [hunk 6]
```diff
@@ -0,0 +1,776 @@
...
+
+    fn get_acked_features(&self) -> u64 {
+        // typical features: 0x17c6effcb
+        // typical acked_features: 0x17000FF8B
+        // typical protocol features: 0x37
+        let mut features = self.device_info.acked_features();
+        // Per the vhost-user spec, MAC is managed by the vhost-user slave.
+        // For ENI, MAC is passed into the hypervisor by the controller.
+        // So disable VIRTIO_NET_F_MAC in the vhost-user slave.
+        features &= !(1 << VIRTIO_NET_F_MAC) as u64;
```
Comment (liubogithub) [inline src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:239]: Is this for non-ENI scenario?

File: src/dragonball/src/api/v1/vmm_action.rs (hunk 1 above)
Comment (justxuewei) [reply src/dragonball/src/api/v1/vmm_action.rs:750]: Yes, upcall is ready for Kata now.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/net.rs (hunk 2 above)
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/vhost/net.rs:16]: This function is used to read an object from memory, but its type is generic without implementing `Debug` trait. No more details are able to retrieve. To get more info, a possible way is to add a `context("extra info here")` when calling this.

Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/vhost/net.rs:16]: UPDATE 1: `anyhow` crate isn't used by Dragonball. So the `context` is unavailable. Then extra info is required to pass from parameters, which I think it isn't a elegant way.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/connection.rs (hunk 3 above)
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/connection.rs:202]: Yes, patch for vhost-user-fs was merged. These lines are able to be removed.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs (hunk 4 above)
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:176]: Removed.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs (hunk 5 above)
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:219]: Fixed.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs (hunk 6 above)
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:235]: Those lines are removed due to no support of ENI for community version.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs [hunk 7]
```diff
@@ -0,0 +1,776 @@
...
+    ) -> VirtioResult<()>
+    where
+        AS: DbsGuestAddressSpace,
+        Q: QueueT + Send + 'static,
+        R: GuestMemoryRegion + Sync + Send + 'static,
+    {
+        info!("{}: try to accept new socket for reconnect...", self.id);
+        match self.listener.try_accept() {
+            Ok(Some((master, _avail_features))) => {
+                // Safe because the `intr_group` has been initialized in VirtioDevice::activate().
```
[the remark this answers is not in the document]
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:261]: Fixed.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs [hunk 8]
```diff
@@ -0,0 +1,776 @@
...
+                warn!("{}: no incoming connection available", self.id);
+                Err(VirtioError::InternalError)
+            }
+        }
+    }
+
+    fn handle_disconnect(&mut self, ops: &mut EventOps) -> VirtioResult<()> {
+        trace!(target: "vhost-net", "{}: VhostUserNetDevice::handle_disconnect()", self.id);
+        self.endpoint.disconnect(ops)
+    }
```
[the remark this answers is not in the document]
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:296]: Fixed.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs [hunk 9]
```diff
@@ -0,0 +1,776 @@
...
+                            "{}: socket disconnected while initializing the connnection: {}",
+                            self.id, err
+                        );
+                    } else {
+                        error!("{}: failed to setup connection: {}", self.id, err);
+                        return Err(VhostError::VhostUserProtocol(err).into());
+                    }
+                }
+                Err(err) => {
+                    error!("{}: failed to setup connection: {}", self.id, err);
```
[the remark this answers is not in the document]
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:316]: `VhostUserProtocol` error branch is to try to reconnect. Others should be one type of errors: print error log and stop performing this function. Sharing one error message is acceptable in my perspective.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs [hunk 10]
```diff
@@ -0,0 +1,776 @@
...
+    R: GuestMemoryRegion + Sync + Send + 'static,
+{
+    fn device(&self) -> MutexGuard<VhostUserNetDevice> {
+        // Do not expect poisoned lock here
+        self.device.lock().unwrap()
+    }
+    fn process_ctrl_request(&mut self) -> VirtioResult<()> {
+        let guard = self.config.lock_guest_memory();
+        let mem = guard.deref();
+        let cvq = self.config.ctrl_queue.as_mut().unwrap();
```
[the remark this answers is not in the document]
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:486]: Added.

**Comment by justxuewei** [comment]
Hi @liubogithub, I've pushed a new commit which resolve issues you mentioned in your code review. PTAL. Thanks!

File: src/dragonball/src/device_manager/vhost_user_net_dev_mgr.rs [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+        ctx: &mut DeviceOpContext,
+    ) -> Result<Box<VhostUserNet<GuestAddressSpaceImpl>>, VirtioError> {
+        match ctx.epoll_mgr.as_ref() {
+            Some(epoll_mgr) => Ok(Box::new(VhostUserNet::new_server(
+                &cfg.sock_path,
+                cfg.guest_mac(),
+                Arc::new(cfg.queue_sizes()),
+                epoll_mgr.clone(),
+            )?)),
+            None => Err(VirtioError::InvalidInput),
```
Comment (studychao) [inline src/dragonball/src/device_manager/vhost_user_net_dev_mgr.rs:124]: how about making this error information more explicit like `NoEpollMgr`?

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs [hunk 12]
```diff
@@ -0,0 +1,776 @@
...
+        if let Some(mac) = guest_mac {
+            // When this feature isn't available, the driver generates a random
+            // MAC address. Otherwise, it should attempt to read the device MAC
+            // address from the config space.
+            avail_features |= 1u64 << VIRTIO_NET_F_MAC;
+            config_space[0..6].copy_from_slice(mac.get_bytes());
+        } else {
+            avail_features &= !(1 << VIRTIO_NET_F_MAC) as u64;
+        }
+        // status: mark link as up
```
Comment (studychao) [inline src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:104]: could the index here (6,7,8,9..16) be a const so that it's clearer which bit it refers to.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs [hunk 13]
```diff
@@ -0,0 +1,776 @@
...
+        config_space[6] = VIRTIO_NET_S_LINK_UP as u8;
+        config_space[7] = 0;
+        // max_virtqueue_pairs: only support one rx/tx pair
+        config_space[8] = (queue_sizes.len() / 2) as u8;
+        config_space[9] = 0;
+        // mtu: 1500 = 1536 - vxlan header?
+        config_space[10] = 220;
```
Comment (studychao) [inline src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:111]: why 220? comment or const?

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs [hunk 14]
```diff
@@ -0,0 +1,776 @@
...
+        config_space[11] = 5;
```
Comment (studychao) [inline src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:112]: ditto

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs [hunk 15]
```diff
@@ -0,0 +1,776 @@
...
+        // speed: 1000Mb
+        config_space[12] = 232;
```
Comment (studychao) [inline src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:114]: ditto

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs [hunk 16]
```diff
@@ -0,0 +1,776 @@
...
+        config_space[13] = 3;
```
Comment (studychao) [inline src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:115]: ditto

**Review by studychao — COMMENTED** [review]
thanks, a few comments

File: src/dragonball/src/dbs_virtio_devices/src/balloon.rs [hunk 17]
```diff
@@ -906,8 +910,10 @@ pub(crate) mod tests {
             let kvm = Kvm::new().unwrap();
             let vm_fd = Arc::new(kvm.create_vm().unwrap());
             let resources = DeviceResources::new();
+            let address_space = create_address_space();
             let config = VirtioDeviceConfig::<Arc<GuestMemoryMmap<()>>>::new(
                 Arc::new(mem),
+                address_space,
```
Comment (studychao) [inline src/dragonball/src/dbs_virtio_devices/src/balloon.rs:916]: nit: could you separate all addreess_space related change to a different commit?

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs [hunk 18]
```diff
@@ -0,0 +1,776 @@
...
+                _ => error!(
+                    "{}: unknown net control request class: 0x{:x}",
+                    NET_DRIVER_NAME, ctrl_hdr.class
+                ),
+            }
+        }
+        Ok(0)
+    }
+}
+impl<AS, Q, R> MutEventSubscriber for VhostUserNetHandler<AS, Q, R>
```
Comment (studychao) [inline src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:536]: blank line

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs [hunk 19]
```diff
@@ -0,0 +1,776 @@
...
+                            "{}: failed to handle control queue request, {:?}",
+                            self.id, err
+                        );
+                    }
+                }
+            }
+            _ => warn!("{}: unknown epoll event slot {}", self.id, events.data()),
+        }
+    }
+    fn init(&mut self, ops: &mut EventOps) {
```
Comment (studychao) [inline src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:572]: blank line?

File: src/dragonball/src/dbs_virtio_devices/src/balloon.rs (hunk 17 above)
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/balloon.rs:916]: I've slipt that changes into a separate commit.

File: src/dragonball/src/device_manager/vhost_user_net_dev_mgr.rs (hunk 11 above)
Comment (justxuewei) [reply src/dragonball/src/device_manager/vhost_user_net_dev_mgr.rs:124]: I think it's great. However, other device managers, e.g. `VirtioNetDeviceMgr`, are still using the `InvalidInput` as well. Shall we file an issue to track this and fix this in another PR?

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs (hunk 12 above)
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:104]: Fixed.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs (hunk 13 above)
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:111]: Fixed.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs (hunk 14 above)
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:112]: This is useless code. They are removed.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs (hunk 16 above)
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:115]: This is useless code and removed.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs (hunk 18 above)
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:536]: Fixed

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs (hunk 19 above)
Comment (justxuewei) [reply src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/net.rs:572]: Fixed.

**Comment by justxuewei** [comment]
Hi @studychao, I've addressed the issues mentioned in your comments. Could you take a look at this again? Thanks!

**Review by studychao — APPROVED** [review]
thanks!

**Review by lifupan — APPROVED** [review]
